File: frontend/src/Services/SubstructureService.ts
import { __BaseService } from "./__BaseService"

class SubstructureService extends __BaseService {
    public async updateSubstructure(
        projectId: string,
        caseId: string,
        dto: Components.Schemas.UpdateSubstructureDto,
    ): Promise<Components.Schemas.SubstructureDto> {
        const res: Components.Schemas.SubstructureDto = await this.put(
            `projects/${projectId}/cases/${caseId}/substructure`,
            { body: dto },
        )
        return res
    }
}

export const GetSubstructureService = () => new SubstructureService()

File: backend/api/Features/Assets/CaseAssets/Substructures/Services/SubstructureService.cs
using api.Context;
using api.Context.Extensions;
using api.Features.Assets.CaseAssets.Substructures.Dtos;
using api.Features.Assets.CaseAssets.Substructures.Dtos.Update;
using api.Features.Cases.GetWithAssets;
using api.Features.Cases.Recalculation;
using api.Features.ProjectIntegrity;
using api.ModelMapping;
using api.Models;

using Microsoft.EntityFrameworkCore;

namespace api.Features.Assets.CaseAssets.Substructures.Services;

public class SubstructureService(
    DcdDbContext context,
    IMapperService mapperService,
    IProjectIntegrityService projectIntegrityService,
    IRecalculationService recalculationService)
{
    public async Task<SubstructureDto> UpdateSubstructure<TDto>(
        Guid projectId,
        Guid caseId,
        Guid substructureId,
        TDto updatedSubstructureDto
    )
        where TDto : BaseUpdateSubstructureDto
    {
        await projectIntegrityService.EntityIsConnectedToProject<Substructure>(projectId, substructureId);

        var existingSubstructure = await context.Substructures.SingleAsync(x => x.Id == substructureId);

        mapperService.MapToEntity(updatedSubstructureDto, existingSubstructure, substructureId);
        existingSubstructure.LastChangedDate = DateTime.UtcNow;

        await context.UpdateCaseModifyTime(caseId);
        await recalculationService.SaveChangesAndRecalculateAsync(caseId);

        return mapperService.MapToDto<Substructure, SubstructureDto>(existingSubstructure, substructureId);
    }
}
